include/boost/dynamic_bitset/detail/dynamic_bitset.hpp

```diff
@@ -85,126 +86,129 @@ namespace boost {
+        return static_cast<std::size_t>(__popcnt(value));
+    }
+
+    template <>
+    BOOST_FORCEINLINE std::size_t popcount<unsigned __int64>(unsigned __int64 value) BOOST_NOEXCEPT
+    {
+#if defined(_M_X64)
+        return static_cast<std::size_t>(__popcnt64(value));
+#else
+        return static_cast<std::size_t>(__popcnt(static_cast< unsigned int >(value))) + static_cast<std::size_t>(__popcnt(static_cast< unsigned int >(value >> 32)));
 #endif
- */
+    }
 
 #elif defined(BOOST_GCC) || defined(__clang__) || (defined(BOOST_INTEL) && defined(__GNUC__))
-     template <>
-     BOOST_FORCEINLINE std::size_t popcount<unsigned int>(unsigned int value) BOOST_NOEXCEPT
-     {
-         return __builtin_popcount(value);
-     }
-
-     template <>
-     BOOST_FORCEINLINE std::size_t popcount<unsigned long>(unsigned long value) BOOST_NOEXCEPT
-     {
-         return __builtin_popcountl(value);
-     }
-
-     template <>
-     BOOST_FORCEINLINE std::size_t popcount<boost::ulong_long_type>(boost::ulong_long_type value) BOOST_NOEXCEPT
-     {
-         return __builtin_popcountll(value);
-     }
+
+    // Note: gcc builtins are implemented by compiler runtime when the target CPU may not support the necessary instructions
+    template <>
+    BOOST_FORCEINLINE std::size_t popcount<unsigned short>(unsigned short value) BOOST_NOEXCEPT
+    {
+        return static_cast<unsigned int>(__builtin_popcount(static_cast<unsigned int>(value)));
+    }
+
+    template <>
+    BOOST_FORCEINLINE std::size_t popcount<unsigned int>(unsigned int value) BOOST_NOEXCEPT
+    {
+        return static_cast<unsigned int>(__builtin_popcount(value));
+    }
+
+    template <>
+    BOOST_FORCEINLINE std::size_t popcount<unsigned long>(unsigned long value) BOOST_NOEXCEPT
+    {
+        return static_cast<unsigned int>(__builtin_popcountl(value));
+    }
+
+    template <>
+    BOOST_FORCEINLINE std::size_t popcount<boost::ulong_long_type>(boost::ulong_long_type value) BOOST_NOEXCEPT
+    {
+        return static_cast<unsigned int>(__builtin_popcountll(value));
+    }
+
 #endif
 
-     // overload for access by blocks
-     //
-     template <typename Iterator, typename ValueType>
-     inline std::size_t do_count(Iterator first, std::size_t length, ValueType,
-                                 value_to_type<access_by_blocks>*)
-     {
-         std::size_t num = 0;
-         while (length){
-             num += popcount<ValueType>(*first);
-             ++first;
-             --length;
-         }
-
-         return num;
-     }
+    // overload for access by blocks
+    //
+    template <typename Iterator, typename ValueType>
+    inline std::size_t do_count(Iterator first, std::size_t length, ValueType,
+                                value_to_type<access_by_blocks>*)
+    {
+        std::size_t num1 = 0u, num2 = 0u;
+        while (length >= 2u) {
+            num1 += popcount<ValueType>(*first);
+            ++first;
+            num2 += popcount<ValueType>(*first);
```

Comment:
It's interesting this improves performance.  Simply by assigning the result two two variables (i.e. registers) that allows them to execute independently even though they are both tied to the advancement of the iterator?  I would have expected to need some sort of hint or intrinsic for that.  Is it the compiler or the CPU doing this?

Reply:
I guess on Intel architecture it's CPU based, after some reading.

Reply:
The iterator is most likely advanced only once per loop iteration. One of the two iterator dereference operations translates to a memory access with a static offset.

The important optimization here is breaking one dependency chain on the accumulator into two, which allows the CPU to execute popcnt and accumulate instructions in parallel for the two dependency chains. Instruction-level parallelism is a common feature of most modern CPUs that support out-of-order execution. Theoretically, a compiler could generate an equivalent code, but I haven't seen a compiler do this.

Reply:
Potentially any loop unroll without side-effects would be eligible for this sort of thing?

Reply:
Yes, most of the time. You may not get any speedup in some cases, e.g. if your algorithm is memory-bound or the loop body becomes too large for the out-of-order window in the CPU. The latter case is somewhat mitigated by instruction reordering done by compiler or by hand.

Reply:
Sounds like an interesting compiler option if you ask me. :)  Thanks for the explanation.